segger-jlink: init at 794a, nrfconnect: init at 4.3.0, nrf-command-line-tools: init at 10.23.2

StarGate01 opened the pull request:
## Description of changes

This PR re-adds some packages which were purged due to their usage of QT4 in #174634 : `segger-jlink`, and two packages which depend on it: `nrfconnect` and `nrf-command-line-tools`.

Unfortunately the SEGGER J-Link tools are only distributed in binary form and have a hard requirement on (the bundled) QT4. I have packaged the bundled QT4 libs as a secondary package to not pollute the global namespace.

Supersedes #252137, #244637, #263487

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

ajs124 commented on pkgs/development/tools/misc/segger-jlink/default.nix:
```diff
@@ -0,0 +1,160 @@
+        curlOpts = "--data accept_license_agreement=accepted";
+      };
+
+  qt4-bundled = stdenv.mkDerivation {
```
Why have this as a second derivation and not just as part of the main one?

StarGate01 replied:
I wanted to ensure that QT4 libs (which are considered outdated) are not contained in /bin or /lib of a top-level package, in order to not pollute the global namespace. Am I mistaken? Should this be included?

h7x4 replied:
I think that makes sense

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/nrfconnect-package-wrongly-removed/33261/2

Mindavi commented:
I think this may be more suitable as an overlay at this point

SuperSandro2000 commented on pkgs/development/tools/misc/nrf-command-line-tools/default.nix:
```diff
@@ -0,0 +1,76 @@
+    mkdir -p $out
+    cp -r * $out
+
+    chmod +x $out/lib/*
```
Executable for libs?

StarGate01 replied:
Hm, I don't know why that is in there. Good catch, I removed it.

SuperSandro2000 commented on pkgs/development/tools/misc/nrf-command-line-tools/default.nix:
```diff
@@ -0,0 +1,76 @@
+
+  version = "10.23.0";
+
+  url = "https://nsscprodmedia.blob.core.windows.net/prod/software-and-other-downloads/desktop-software/nrf-command-line-tools/sw/versions-${lib.versions.major version}-x-x/${lib.versions.major version}-${lib.versions.minor version}-${lib.versions.patch version}/nrf-command-line-tools-${lib.versions.major version}.${lib.versions.minor version}.${lib.versions.patch version}_${platform.name}.tar.gz ...
```
We should deduplicate the code by using a let in here.

StarGate01 replied:
I am unsure how this could be deduplicated, since the fragments are separated by different characters.

MGlolenstine replied:
I think you could create a single string with one separator and then call [`replaceStrings`](https://noogle.dev/?selected=%22builtins.replaceStrings%22&term=%22replace%22) on it to switch the delimiters.

h7x4 replied:
Maybe something like this then?

```suggestion
  url = let
    versionWithDashes = builtins.replaceStrings ["."] ["-"] version;
  in "https://nsscprodmedia.blob.core.windows.net/prod/software-and-other-downloads/desktop-software/nrf-command-line-tools/sw/versions-${lib.versions.major version}-x-x/${versionWithDashes}/nrf-command-line-tools-${version}_${platform.name}.tar.gz";
```

SuperSandro2000 commented on pkgs/development/tools/misc/nrf-command-line-tools/default.nix:
```diff
@@ -0,0 +1,76 @@
+  supported = {
+    x86_64-linux = {
+      name = "linux-amd64";
+      sha256 = "1b69272c704f4526b8560fe3c64b92769626261c2a285258bde550d3dd218873";
+    };
+    aarch64-linux = {
+      name = "linux-arm64";
+      sha256 = "bcad32f9340c60374589e6613aa773dd9255e0cea7c50b9bd83ee47ca170a536";
+    };
+    armv7l-linux = {
+      name = "linux-armhf";
+      sha256 = "7b217b434e675dba4ae3a6f611fb9e15835bbe106fd1a4a4052ae68954111a2e";
```
```suggestion
      hash = "1b69272c704f4526b8560fe3c64b92769626261c2a285258bde550d3dd218873";
    };
    aarch64-linux = {
      name = "linux-arm64";
      hash = "bcad32f9340c60374589e6613aa773dd9255e0cea7c50b9bd83ee47ca170a536";
    };
    armv7l-linux = {
      name = "linux-armhf";
      hash = "7b217b434e675dba4ae3a6f611fb9e15835bbe106fd1a4a4052ae68954111a2e";
```
we should also use sri hashes here

SuperSandro2000 commented on pkgs/development/tools/misc/segger-jlink/default.nix:
```diff
@@ -0,0 +1,160 @@
+  pname = "segger-jlink";
+  inherit src version;
+
+  # Currently blocked by patchelf bug
+  # https://github.com/NixOS/patchelf/pull/275
+  #runtimeDependencies = [ udev ];
```
Isn't this fixed in patchelf 0.13?

StarGate01 replied:
I think the bug is somewhat unrelated, Udev is loaded late in the runtime of the program (I guess via dlopen()). Patchelf does not pick up this dependency, and adding `udev` to `runtimeDependencies` does not work either. So I kept the current approach. Otherwise this happens:

```
Connecting to J-Link via USB...FAILED: USBBULK: Failed to load libudev.so. Needed for identification of J-Links connected via USB
Cannot connect to J-Link.
```

liarokapisv replied:
If I recall correctly `runtimeDependencies` only looks for executables. You want `appendRunpaths`

StarGate01 replied:
Thanks you, added. Learned something new today :)

pinpox commented:
@StarGate01 Are you still working on this? I'm waiting for this package.

StarGate01 commented:
@pinpox Yes I am , I'll implement the suggestions made by @SuperSandro2000 , however I don't really agree with  @Mindavi - I would want to keep this as a package instead of a (somewhat cumbersome) overlay.

liarokapisv commented:
The J-Links are an industry standard,  I believe that the package is definitely worth keeping. I don't think that leaving the bundled qt4 libraries as is, is that huge of a deal in this case.

StarGate01 commented:
I have implemented most of the suggestions by @SuperSandro2000 , and also updated all three packages to the latest versions.

liarokapisv commented:
@StarGate01  I think it is useful to also provide desktop files for the graphical applications like `JLinkRTTViewer`, `JMem` etc.
I have a working reference implementation in my [own pull-request ](https://github.com/NixOS/nixpkgs/pull/263487)(which I will probably close in favor of this one) if it is deemed useful.

StarGate01 commented:
@liarokapisv Good idea - is the list at https://github.com/liarokapisv/nixpkgs/blob/1f6683d53373179b761055239616c7e65a0a71fc/pkgs/by-name/jl/jlink-software-tools/package.nix#L95 the complete list of all apps which should have a dektop icon?

liarokapisv commented:
I think so. It was taken from the AUR package. I double-checked it but I could be missing something.

StarGate01 commented:
Okay. Some entries like `JLink` are CLI tools, would it be a sensible to add a dektop entry for those CLI-only tools?

liarokapisv commented:
I went through all entries, and it seems the following are CLI-only tools:

- JLink
- JRun
- JTagLoad
- JLinkGUIServer
- JLinkRTTClient
- JLinkRTTLogger
- JLinkSTM32

The [segger wiki](https://wiki.segger.com/J-Link_GUI_Software#List_of_J-Link_GUI_applications) seems to have a list of GUI applications.

StarGate01 commented:
@liarokapisv I added symlinks for all executables, and also added the missing runtime files for a few specific applications.

I added code to generate and install the desktop entries for the GUI applications you listed, but it does not seem to work yet. Do you have an idea why the icons won't show?

liarokapisv commented:
@StarGate01 The desktop entries seem to be working fine for me on both Nixos and Standalone:

![](https://github.com/NixOS/nixpkgs/assets/19633626/c11f443f-e309-4ee2-999c-b598e0672a25|width=20)

StarGate01 commented:
@liarokapisv You are correct, I had my system misconfigured. I think this solution is fine, what do you think?

liarokapisv commented:
Seems fine to me!

StarGate01 commented:
@SuperSandro2000 any objections on this PR?

h7x4 reviewed (commented):
Mostly non-blocking nits

StarGate01 commented on pkgs/development/tools/misc/segger-jlink/default.nix:
```diff
@@ -0,0 +1,202 @@
+    mv libjlinkarm.so* $out/lib
+    # This library is opened via dlopen at runtime
+    for libr in $out/lib/*; do
+      ln -s $libr $out/bin
```
The original implementation in the dropped package was written by me as well.

IIRC, the issue was that the JLink executable use dlopen() at runtime to load the library. Somehow, this did not work when the library was placed in the lib/ directory, maybe due to the executable using a relative path, or the lib/ directory not being added to the PATH of the the executable.

I will investigate and report back.

StarGate01 replied:
Okay, It turns out that the J-Link tools do indeed load `libjlinkarm` late at runtime. Adding this dependency using `patchelf --add-needed libjlinkarm.so` did not work, and neitehr did adding `$ORIGIN/../lib` to `appendRunpaths`. strace show that the library is also being loaded from `opt/`:

```
openat(AT_FDCWD, "/nix/store/p587y1wxllcys9xk8xpgydf4g2kal6ba-segger-jlink-794a/bin/../opt/libjlinkarm.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/nix/store/p587y1wxllcys9xk8xpgydf4g2kal6ba-segger-jlink-794a/opt/libjlinkarm.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/opt/SEGGER/JLink/libjlinkarm.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/opt/SEGGER/JLink/libjlinkarm.so.5", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
```

I modified the code to link the libraries into `opt/`  instead of `bin/`, which I think is a cleaner solution as it wont pollute the PATH that way.

StarGate01 commented:
Thank you @h7x4 for your comments, I have implemented / commented your suggestions.

RaitoBezarius commented:
Can the Qt 4 known vulnerabilities be copied over this package (if not already)? Thank you!

StarGate01 commented:
@RaitoBezarius  What exactly are you referring to? Code patches? Documentation? Do you have a link?

RaitoBezarius commented:
> @RaitoBezarius What exactly are you referring to? Code patches? Documentation? Do you have a link?

Qt 4 had `knownVulnerabilities` field regarding known CVEs, as this is using a bundled Qt 4, it should replicate this metadata field for transparency.

teburd commented:
It would be great to be able to use jlink again without copying/pasting my own overlays for it, I have an expression that adds systemview as well but was waiting on this to happen before sending a PR for it

StarGate01 commented:
@teburd I have a PR for SystemView which is rebased onto this one : #214195

RaitoBezarius commented on pkgs/development/tools/misc/segger-jlink/default.nix:
```diff
@@ -0,0 +1,207 @@
+      license = licenses.lgpl21;
+      maintainers = with maintainers; [ stargate01 ];
+      knownVulnerabilities = [
+        "This bundled version of Qt 4 has reached its end of life after 2015. See https://github.com/NixOS/nixpkgs/pull/174634."
```
```suggestion
        "This bundled version of Qt 4 has reached its end of life after 2015. See https://github.com/NixOS/nixpkgs/pull/174634"
```

I would also add all relevant CVEs in here: https://www.cvedetails.com/vulnerability-list/vendor_id-6363/product_id-10758/version_id-660427/QT-QT-4.8.0.html.

RaitoBezarius replied:
Thank you!

maxhbr commented on pkgs/tools/bluetooth/nrfconnect/default.nix:
```diff
@@ -0,0 +1,44 @@
+{ lib
+, fetchurl
+, appimageTools
+}:
```
this is missing `segger-jlink` as an input.

maxhbr replied:
ah, sorry, the reason was my way of using it. If segger-jlink is in pkgs, everything is fine.

maxhbr reviewed (changes requested):
nrfconnect/default.nix is missing `segger-jlink` as an input

maxhbr reviewed (approved):
Works for me :+1:

h7x4 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -19700,8 +19700,12 @@ with pkgs;
     sdk = true;
   };
 
+  nrf-command-line-tools = callPackage ../development/tools/misc/nrf-command-line-tools { };
```
I think these maybe should go to `pkgs/by-name` now.

h7x4 replied:
Yeah, I think so. There's [RFC 140](https://github.com/NixOS/rfcs/pull/140) that details the move, but there's a shorter description at [`pkgs/by-name/README.md`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/README.md). I think after #275539, this should've become mandatory for new packages, but the pipeline might not have triggered on it since the last push was before this was merged.

StarGate01 replied:
Okay, thank you. I have moved the packages to their new location.

StarGate01 committed:
segger-jlink: init at 794a

StarGate01 committed:
nrfconnect: init at 4.3.0

StarGate01 committed:
nrf-command-line-tools: init at 10.23.2

teburd commented:
Works for me, and I'm using it in a flake today in a very janky way. Would really much prefer this in tree

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/3395

Mindavi commented:
Okay, there seems to be a need for this, and hydra won't help building this anyway. So I guess we can carry this package for now. But I'm not super thrilled about carrying such an old qt, it'll inevitably break at some point, with a compiler update or so. Anyhow, it won't burden hydra much.

h7x4 commented:
Maybe it would be possible to split the package? I suspect that most things depending on segger jlink only depends on the libraries and/or cli tools, not the GUI stuff. Some kind of `segger-jlink-headless`?

Mindavi commented:
QT also contains quite some libraries, it seems unlikely only the GUI parts were used and none of the non-gui parts.

StarGate01 commented:
Interesting idea, we would have to figure out which of the tools actually use QT. This might also change with future updates. I am unsure if there is a usecase for a "reduced" JLink tool suite - even mundane things like a license popup use QT4.

emilazy commented:
I’ve opened https://github.com/NixOS/nixpkgs/pull/353880 to revert this, per https://github.com/NixOS/nixpkgs/pull/214195#issuecomment-2457822890.
